### PR TITLE
[WIP] Add tutorial for MPdist

### PR DESCRIPTION
**Pull request by asifmallik**
This completes #290 

Just noticed that I incorrectly capitalized d in MPdist so I need to correct that. Another thing is to decide whether to include a hierarchical clustering for euclidean distance. I was not able to replicate the figure from the paper so far. I am finalizing a notebook currently which showcases multiple attempts to replicate the figure and will post it in Issues.

# Pull Request Checklist

- [x] Fork, clone, and checkout the newest version of the code
- [x] Create a new branch
- [x] Make necessary code changes
- [x] Install `black` (i.e., `python -m pip install black` or `conda install -c conda-forge black`)
- [x] Install `flake8` (i.e., `python -m pip install flake8` or `conda install -c conda-forge flake8`)
- [x] Install `pytest-cov` (i.e., `python -m pip install pytest-cov` or `conda install -c conda-forge pytest-cov`)
- [x] Run `black .` in the root stumpy directory
- [x] Run `flake8 .` in the root stumpy directory
- [ ] Run `./setup.sh && ./test.sh` in the root stumpy directory
- [x] Reference a Github issue (and create one if one doesn't already exist)

Skipped `./test.sh`because this is a documentation only pull request

### CODE REVIEW

File: docs/Tutorial_MPDist.ipynb [hunk 1]
```diff
@@ -0,0 +1,309 @@
+{
```
Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <p>I think it's <em>hierarchical </em>clustering, not entirely sure though so double check this.</p><p> </p><p>You can capitalize "Euclidean".</p><p> </p><p>Replace "MPdist" with "MPDist" in the title (or change everywhere else to "MPdist", whichever is more convenient).</p><p> </p><blockquote>More concretely. given a threshold p and window length L, the MPdist between two time series is the smallest distance such that p * 100% of their windows of size L have a lower euclidean distance to some window of equal size in the other .&nbsp;</blockquote><p> </p><p>Maybe you could rephrase this, I wasn't sure what this meant on first glance. Maybe say p% instead of p * 100%? Maybe rephrase to something like: "... the MPDist between two time series is the smallest pairwise Euclidean distance such that it is higher in value than p% of pairwise Euclidean distances", or some variation of this. It's a tough definition to summarize...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <p>You could add labels to the axes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    plt.rcParams["figure.figsize"] = [20, 4]  # width, height
</pre><p>You can replace these with <code>plt.style.use('stumpy.mplstyle')</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #19.    mp_dist = stumpy.mpdist(t_1, t_2, 30, percentage=0.05)
</pre><p>Might consider storing <code>m = 30</code>  and <code>percentage = 0.05</code>  since you use this multiple times.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <p>Consider replacing the for loop with this:</p><pre class="ql-syntax" spellcheck="false">data = loadmat('MaryBethAnhLisa_data.mat')['XSsY4']
dfs = {data[1][i][0]:pd.DataFrame(data[0][i].flatten()) for i in range(data[0].shape[0])}
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <p>This is more of an aesthetic suggestion, consider using different colors for the line charts.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>To be consistent with the paper, let's use "MPdist" everywhere please</p>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>I think the for-loop is easier to read</p>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>It might actually be useful to align "Mary" with the start of "Maryanne" and "Anne" with the end of "Maryanne", etc so it is easier for people to visually see that they are similar</p>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>Instead of <code>p</code>  or <code>L</code> , let's use the same variables that are available in the <code>mpdist</code>  function</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>Do you really mean "self-join" or do you mean "repeated"? In matrix profiles, "self-join" has a specific meaning so we don't want to overload the that term. Also, why repeat it 10 times? Why isn't 3 times sufficient?</p><p> </p><p>Also, instead of</p><p> </p><pre class="ql-syntax" spellcheck="false">base = np.concatenate([motif] * 10)&nbsp;
</pre><p> </p><p>Maybe you can do:</p><p> </p><pre class="ql-syntax" spellcheck="false">base = np.tile(motif, 10)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>Let's replace "female" with, maybe" "American"? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>I think it would be good to add a little bit more of an explanation of how the eye movement relates to the plot</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>I don't think we need to say this here as long as we understand that what we are doing is "correct" and the outcome is as expected</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

**Comment by seanlaw** [comment]
@asifmallik So far, so good! Good work. I like where this is going

File: docs/Tutorial_MPDist.ipynb (hunk 1 above)
Comment (alvii147) [reply docs/Tutorial_MPDist.ipynb:1]: Yea I agree, the dictionary comprehension is far from readable. Maybe I was just challenging myself to write a one liner.....

In any case, the final version should look like neither since we'll upload the datasets to Zenodo. Maybe you could do the data wrangling locally and THEN upload the cleaned dataset to Zenodo, so that on the final tutorial, it's just a matter of indexing a single dimension.

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Fixed the spelling and captilization errors - thank you for pointing them out! As for rephrasing the paragraph, I tried doing a better job by referencing join matrix profile - do you think I should define that as well or maybe link back to the paper?</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>In this case the sequences are randomly generated so they don't really mean anything, any suggestions on how to label them?</p><p> </p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Hmm, although the values are same, they are used in two completely different contexts so I don't think they necessarily need to be the same. Moreover, it's probably more easier to see what the parameters are instead of having to refer back if the reader forgets. </p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Sounds good</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>I am not sure whether self-join means self-concatenation for time series but I will change it to repeated just to disambiguate with the way they use it in matrix profile context. Changed it to tile as well. 10 is unnecessarily high probably - switched to 50 instead and remove the axis limits</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Hmm, I can remove it but it might confuse some readers if they see that the two do not match</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Fixed!</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Yeah, cleaning it first sounds like a good idea, I will fix this later</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>I vaguely remember reading eye movement somewhere but I am not able to find it in the paper. Any idea where this is explained?</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: Good catch - thank you!

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: Is using Zenodo something I can do for this something I can do? @seanlaw

**Comment by asifmallik** [comment]
@seanlaw @alvii147 Made some changes and incorporated some of your suggestions and corrections. Still have a bit left though

File: docs/Tutorial_MPDist.ipynb (hunk 1 above)
Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>I think it's still okay to say "Time" on the x-axis and "Random Value" on the y-axis</p>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>@asifmalik I can help you with that. I'm thinking that we'll just create 6 separate files (one for each name)?</p>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>I don't know if this is the same but here's what I've read before in another similar dataset: https://stumpy.readthedocs.io/en/latest/Tutorial_Consensus_Motif.html?highlight=eye#Loading-the-Eye-tracking-(EOG)-Dataset</p>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Okay, then I would omit "a mistake in the paper". I think that we can explain the difference without sounding like we are assigning fault. Does that make sense?</p>

**Comment by seanlaw** [comment]
@asifmallik No problem! I value quality over quantity and what you've done so far is really taking shape

File: docs/Tutorial_MPDist.ipynb [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
```
Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    plt.style.use('stumpy.mplstyle')
</pre><p> </p><p>Can you replace this line with:</p><p> </p><pre class="ql-syntax" spellcheck="false">plt.style.use('https://raw.githubusercontent.com/TDAmeritrade/stumpy/main/docs/stumpy.mplstyle')
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

File: docs/Tutorial_MPDist.ipynb (hunk 1 above)
Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: Yes, makes sense!

**Comment by asifmallik** [comment]
@seanlaw ready for a more thorough review

File: docs/Tutorial_MPDist.ipynb (hunk 1 above)
Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: They are aligned now but this required some extra code for writing offset indexes

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: I wasn't able to find anything about it being eye-tracking or EOG, but the language in the paper seemed to me to indicate that it was pen position so I went with that

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: Got it, it's fixed now

File: docs/Tutorial_MPDist.ipynb [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
```
Comment (asifmallik) [inline docs/Tutorial_MPDist.ipynb:1]: <p>*of their subsequences</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Comment by seanlaw** [comment]
> @seanlaw ready for a more thorough review

@asifmallik I will try to find some time to review it. Thank you

File: docs/Tutorial_MPDist.ipynb (hunk 3 above)
Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Instead of "determining whether most" maybe it should be "determining whether a limited subset of subsequences - as parameterized by a threshold - are similar"</p>

Comment (seanlaw) [reply docs/Tutorial_MPDist.ipynb:1]: <p> </p><blockquote>For instance, if two time series is made up of the same repeating subsequences of window length <code>m</code>, then they would MPdist of 0 (if window size for MPdist is set to <code>m</code>), even if they are phase shifted. On the other hand, the Euclidean distance would be non-zero as long as they are phase shifted. </blockquote><p>I find this sentence to be very confusing/abstract as it is not anchored to anything and expects the user to already have some prior knowledge</p>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>Please remove the comma as it is unnessary</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #19.    mplstyle_url = 'https://raw.githubusercontent.com/TDAmeritrade/stumpy/main/docs/stumpy.mplstyle'
</pre><p> </p><p>Please replace with </p><p> </p><pre class="ql-syntax" spellcheck="false">plt.style.use("https://raw.githubusercontent.com/TDAmeritrade/stumpy/main/docs/stumpy.mplstyle")
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    t_1 = base[:50]
</pre><p> </p><p>Please use <code>T_A</code>  and <code>T_B</code>  to refer to time series A and time series B</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    plt.style.use(mplstyle_url)
</pre><p>Please remove the above line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>I wonder if it would make more sense to have the long name first and then followed by the short parts</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>It feels like you are saying a lot here but you aren't showing the evidence for it until later. It is okay to make a statement and show the evidence inline.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (seanlaw) [inline docs/Tutorial_MPDist.ipynb:1]: <p>What do the colors in the dendrogram mean?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

**Comment by seanlaw** [comment]
@asifmallik I really like the story with both the Euclidean distance vs MPdist. It "feels" complete. 

I was wondering if it would be possible to start with a simple example that only focuses on using MPdist to compare just two time series first. This way the focus is purely on MPdist and its output. After we've done through MPdist then we can talk about the name example. Otherwise, the dendrogram work might overshadow the point of this tutorial and that is to learn about MPdist and why it is useful and how it works.

**Comment by asifmallik** [comment]
> @asifmallik I really like the story with both the Euclidean distance vs MPdist. It "feels" complete.
> 
> I was wondering if it would be possible to start with a simple example that only focuses on using MPdist to compare just two time series first. This way the focus is purely on MPdist and its output. After we've done through MPdist then we can talk about the name example. Otherwise, the dendrogram work might overshadow the point of this tutorial and that is to learn about MPdist and why it is useful and how it works.

Do you mean I should replace the current introduction which starts with two randomly generated time series with the time series for two of the names instead?

**Comment by seanlaw** [comment]
> Do you mean I should replace the current introduction which starts with two randomly generated time series with the time series for two of the names instead?

I would like to replace the random data example and use the data from Figure 1 (and maybe Figure 2) and also motivate and explain the pitfalls of other measures/methods by following what the authors covered/discussed in the introduction.

File: docs/Tutorial_MPDist.ipynb (hunk 3 above)
Comment (alvii147) [reply docs/Tutorial_MPDist.ipynb:1]: <p><em>For instance, if two time series <s>is</s> </em><strong><em>are</em></strong><em> made up of the same repeating subsequences of window length </em><code><em>m</em></code><em>, then they would </em><strong><em>have an</em></strong><em> MPdist of 0...</em></p><p> </p><p>I agree it's a bit difficult to follow this definition, although it's tough to come up with an alternative phrasing. Maybe something like: <em>For instance, two time series that are made up of the same periodic subsequences, but are phase shifted, their MPdist would be zero, while their Euclidean distance would be non-zero.</em></p><p> </p><p>I understand the significance of <code>m</code> , but maybe in this exact definition you can remove the mention of it? Idk it seems cleaner for the sake of explaining.</p>

Comment (alvii147) [reply docs/Tutorial_MPDist.ipynb:1]: <p>You could just move the comma forward so it makes more sense: <em>We begin by importing STUMPY </em><strong><em>and</em></strong><em> some other relevant packages for loading</em><strong><em>,</em></strong><em> and analyzing the data.</em></p>

File: docs/Tutorial_MPDist.ipynb (hunk 1 above)
Comment (alvii147) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Oh true. Maybe you could do <code>m = 30</code>  twice in two different code blocks, for the sake of cleanliness. Hardcoded values are often confusing to me without context.</p>

File: docs/Tutorial_MPDist.ipynb (hunk 3 above)
Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <p>It would look nicer if the repeated subsequences were highlighted, or more visible somehow.</p><p> </p><p>Try this (or something similar):</p><p> </p><pre class="ql-syntax" spellcheck="false">from matplotlib.patches import Rectangle
</pre><p> </p><pre class="ql-syntax" spellcheck="false">rect = Rectangle((0, 0), 20, 50, facecolor='lightgrey')
axs[0].add_patch(rect)
axs[0].plot(np.arange(20), t_1[:20], color='aquamarine', linewidth=10, alpha=0.5)

rect = Rectangle((5, 0), 20, 50, facecolor='lightgrey')
axs[1].add_patch(rect)
axs[1].plot(np.arange(5, 25), t_2[5:25], color='aquamarine', linewidth=10, alpha=0.5)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (alvii147) [inline docs/Tutorial_MPDist.ipynb:1]: <p>i and j look better as <code>i</code>  and <code>j</code> . </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/433/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

[the remark this answers is not in the document]
Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: Oh, I see, thank you, I didn't know this

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: Reading the docs, I am not exactly quite sure what it means, seems like they color it something depending on whether the distance of the child nodes exceeds a certain threshold (default is set to be 70% of max distance). This doesn't seem particularly relevant for this case so I can just set `link_color_func` to be a function that returns the same color for every cluster.

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Yes, I think that's a good idea too, changing the ordering</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb:1]: <p>Removed</p>

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb]: Resolved

Comment (asifmallik) [reply docs/Tutorial_MPDist.ipynb]: Resolved

**Comment by seanlaw** [comment]
@asifmallik Any updates on this?
